tests/model_serving/model_server/maas_billing/utils.py

```diff
@@ -249,20 +262,198 @@ def verify_chat_completions(
 
     LOGGER.info(f"{log_prefix}: POST {model_url} -> HTTP {response.status_code}")
 
-    assert response.status_code == 200, (
-        f"/v1/chat/completions failed: HTTP {response.status_code} response={response.text[:200]} (url={model_url})"
+    assert response.status_code in expected_status_codes, (
+        f"/v1/chat/completions failed: HTTP {response.status_code} "
+        f"response={response.text[:200]} (url={model_url}), "
+        f"expected one of {expected_status_codes}"
+    )
+
+    if response.status_code == 200:
+        response_body = response.json()
+        completions_choices = response_body.get("choices", [])
+        assert isinstance(completions_choices, list) and completions_choices, (
+            "'choices' field missing or empty in /v1/chat/completions response"
+        )
+
+        first_choice = completions_choices[0]
+        message_section = first_choice.get("message", {}) or {}
+        content_text = message_section.get("content") or first_choice.get("text", "")
+
+        assert isinstance(content_text, str) and content_text.strip(), (
+            "First choice in /v1/chat/completions response has no text content"
+        )
+
+    return response
+
+
+def assert_mixed_200_and_429(
+    *,
+    actor_label: str,
+    status_codes_list: List[int],
+    context: str,
+    require_429: bool = True,
+) -> None:
+    assert status_codes_list, f"{actor_label}: no responses in {context}"
+
+    allowed_status_codes = {200, 429}
+    unexpected_status_codes = [code for code in status_codes_list if code not in allowed_status_codes]
+
+    assert not unexpected_status_codes, (
+        f"{actor_label}: unexpected HTTP status codes in {context}: {unexpected_status_codes}. "
+        f"Full sequence={status_codes_list}. "
+        "Likely wrong endpoint/route (404) or server/gateway issue (5xx)."
     )
 
-    response_body = response.json()
-    completions_choices = response_body.get("choices", [])
-    assert isinstance(completions_choices, list) and completions_choices, (
-        "'choices' field missing or empty in /v1/chat/completions response"
+    assert status_codes_list[0] == 200, (
+        f"{actor_label}: expected first status 200 in {context}, got {status_codes_list[0]} "
+        f"(status_codes={status_codes_list})"
     )
```

Comment:
How are you validating that till rate limit is reached you are getting 200's and only after rate limit is reached you are getting 429?